File: toolkit-stocks/src/domain/intrinsic_value/intrinsic_value.py
import math


def free_cash_flows(cash_flow):
    dates = []
    values = []
    for i in range(len(cash_flow)):
        if i + 1 < len(cash_flow) and cash_flow[i]["fiscal_date"] <= cash_flow[i + 1]["fiscal_date"]:
            raise ValueError("Cash Flow statements corrupted")
        value = cash_flow[i]["free_cash_flow"]
        if isinstance(value, (int, float)) and not math.isnan(value):
            dates.append(cash_flow[i]["fiscal_date"])
            values.append(value)
    return {"dates": dates, "values": values}


def annual_growth_rate(fcfs):
    periods = len(fcfs)
    return (fcfs[0] / fcfs[-1]) ** (1 / periods) - 1


def normalized_free_cash_flow(fcfs):
    return sum(fcfs) / len(fcfs)


def future_free_cash_flows(fcf, length, dagr):
    holder = []
    current = fcf
    for _ in range(length):
        current = current * (1 + dagr)
        holder.append(current)
    return holder


def price_to_free_cash_flow(market_cap, fcf):
    return market_cap / fcf


def discounted_future_free_cash_flows(ffcfs, length, disc):
    holder = []
    for exp in range(1, length + 1):
        index = exp - 1
        holder.append(ffcfs[index] / disc ** exp)
    return holder


def calculate_intrinsic_value(
    market_cap,
    stock_price,
    cash_and_equivalents,
    share_issued,
    cash_flow,
    growth_rate_discount,
    future_discount_rate,
    safety_margin,
    projection,
):
    grd_decimal = 1 - growth_rate_discount / 100
    fdr_decimal = 1 + future_discount_rate / 100
    sm_decimal = 1 - safety_margin / 100

    result = free_cash_flows(cash_flow)
    fcfs = result["values"][:projection]
    projection = len(fcfs)

    agr = annual_growth_rate(fcfs)
    dagr = agr * grd_decimal

    nfcf = normalized_free_cash_flow(fcfs)
    ffcfs = future_free_cash_flows(nfcf, projection, dagr)

    pfcf = price_to_free_cash_flow(market_cap, fcfs[0])
    ter = ffcfs[-1] * pfcf

    dffcfs = discounted_future_free_cash_flows(ffcfs, projection, fdr_decimal)
    dffcf_sum = sum(dffcfs)
    dter = ter / (fdr_decimal ** projection)
    pv = dffcf_sum + dter

    iv = pv + cash_and_equivalents

    ivps = (iv * sm_decimal) / share_issued
    ivpspd = ((ivps - stock_price) / stock_price) * 100

    return {
        "stock_price": stock_price,
        "market_cap": [
            market_cap,
        ],
        "projection_horizon": projection,
        "annual_growth_rate": [
            agr * 100,
        ],
        "intrinsic_value": [
            iv,
        ],
        "intrinsic_value_per_share": [
            ivps,
        ],
        "iv_upside": [
            ivpspd,
        ],
    }

File: web/server/calculations/intrinsicValue.js
import { roundAnyNumber } from "#shared/functions/numeric";

function calculateNetIncome(income, dividends, shares) {
	const div = dividends || 0;
	return (income - div) / shares + 1;
}


/**
 * intrinsic value without margin of safety.
 * returns dictionary with analysis.
 * @param {number} price newest (closing) price
 * @param {number} marketCap newest
 * @param {number} cashAndEquivalents from newest balance sheet
 * @param {number} shareIssued from newest balance sheet
 * @param {object[]} cashFlow list of statements (newest first)
 * @param {number} growthRateDiscount growth rate
 * @param {number} futureDiscountRate Discount for future free cash flow and terminal value.
 * @returns {object}
 */
export default function calculateIntrinsicValue(
	marketCap,
	stockPrice,
	cashAndEquivalents,
	shareIssued,
	cashFlow,
	growthRateDiscount,
	futureDiscountRate,
	safetyMargin,
) {
	const grdDecimal = 1 - growthRateDiscount / 100;
	const fdrDecimal = 1 + futureDiscountRate / 100;
	const smDecimal = 1 - safetyMargin / 100

	/** list of dates */
	// const dates = cashFlow.map((item) => item["fiscal_date"].split("-")[0]);

	/** free cash flows from newest to oldest */
	const { dates, values: fcfs } = freeCashFlows(cashFlow);
	const projection = fcfs.length;

	const agr = annualGrowthRate(fcfs);
	const dagr = agr * grdDecimal;

	const nfcf = normalizedFreeCashFlow(fcfs);
	const ffcfs = futureFreeCashFlows(nfcf, projection, dagr);

	const pfcf = priceToFreeCashFlow(marketCap, fcfs[0]);
	const ter = ffcfs.at(-1) * pfcf;

	const dffcfs = discountedFutureFreeCashFlows(ffcfs, projection, fdrDecimal);
	const dffcfSum = dffcfs.reduce((acc, v) => acc + v, 0);
	const dter = ter / (fdrDecimal ** projection);
	const pv = dffcfSum + dter;

	const iv = pv + cashAndEquivalents;

	const ivps = (iv * smDecimal) / shareIssued;
	// intrinsic value per share percentage difference
	const ivpspd = ((ivps - stockPrice) / stockPrice) * 100

	return {
		"stock_price": stockPrice,
		"market_cap": [
			marketCap,
			roundAnyNumber(marketCap, 2)
		],
		"projection_horizon": projection,
		"annual_growth_rate": [
			agr * 100,
			roundAnyNumber(agr * 100, 1)
		],
		// "free_cash_flows": Object.fromEntries(
		// 	fcfs.map((val, i) => [dates[i], val])
		// ),
		// "future_free_cash_flows": ffcfs.reduce((acc, val, i) => {
		// 	acc[Number(dates[0]) + i + 1] = val;
		// 	return acc;
		// }, {}),
		// "discounted_future_free_cash_flows": dffcfs.reduce(
		// 	(acc, val, i) => {
		// 		acc[Number(dates[0]) + i + 1] = val;
		// 		return acc;
		// 	}, {}),
		"intrinsic_value": [
			iv,
			roundAnyNumber(iv, 2)
		],
		"intrinsic_value_per_share": [
			ivps,
			roundAnyNumber(ivps, 2)
		],
		"iv_upside": [
			ivpspd,
			`${ivpspd >= 0 ? "+" : ""}${roundAnyNumber(ivpspd, 1)}%`
		]
	};
}

/**
 * list of free cash flows, starting from newest
 * @param {object[]} cashFlow
 * @param {number} index
 * @param {{dates: string[], values: number[]}} holder
 * @returns {{dates: string[], values: number[]}}
 */
function freeCashFlows(cashFlow, index = 0, holder = { dates: [], values: [] }) {
	if (index >= cashFlow.length) return holder;

	// check if statements are from newest to oldest
	const date = (i) => cashFlow[i]["fiscal_date"];

	if (index + 1 < cashFlow.length &&
		date(index) <= date(index + 1)
	) {
		throw createError({ statusCode: 400, statusMessage: 'Cash Flow statements corrupted' })
	};

	const value = cashFlow[index]["free_cash_flow"];

	if (typeof value === "number" && !isNaN(value)) {
		holder.dates.push(date(index));
		holder.values.push(value);
	}

	return freeCashFlows(cashFlow, index + 1, holder);
};


/**
 * @param {number[]} fcfs
 * @returns {number}
 */
function annualGrowthRate(fcfs) {
	const periods = fcfs.length;
	const first = fcfs[0];
	const last = fcfs.at(-1);

	// negative/zero → no growth projection, conservative 0
	if (last === 0 || first / last < 0) return 0.0001; 

	return (first / last) ** (1 / periods) - 1;
};

/**
 * @param {number[]} fcfs
 * @returns {number}
 */
function normalizedFreeCashFlow(fcfs) {
	const sum = fcfs.reduce((acc, v) => acc + v, 0);
	return sum / fcfs.length;
}

/**
 * @param {number} fcf normalized free cash flow
 * @param {number} lfcfs length of the fcf array
 * @param {number} dagr discounted annual growth rate of fcf
 * @param {number[]} holder hold values in array to return
 * @returns {number[]}
 */
function futureFreeCashFlows(fcf, lfcfs, dagr, holder = []) {
	if (lfcfs <= 0) return holder;

	const dfcf = fcf * (1 + dagr);
	return futureFreeCashFlows(
		dfcf,
		lfcfs - 1,
		dagr,
		[...holder, dfcf]
	);
}

/**
 * TODO FCF newest value - must be average across industry
 * @param {number} marketCap market cap
 * @param {number} fcf newest free cash flow
 */
function priceToFreeCashFlow(marketCap, fcf) {
	return marketCap / fcf;
};

/**
 * @param {number[]} ffcfs future free cash flows
 * @param {number} length length of ffcfs
 * @param {number} disc discount
 * @param {number} exp exponent + also works as an index
 * @param {number[]} holder
 * @returns {number[]}
 */
function discountedFutureFreeCashFlows(
	ffcfs,
	length,
	disc,
	exp = 1,
	holder = []
) {
	if (length <= 0) return holder;

	const index = exp - 1;
	return discountedFutureFreeCashFlows(
		ffcfs,
		length - 1,
		disc,
		exp + 1,
		[...holder, ffcfs[index] / disc ** exp]
	);
};
